### PR TITLE
[inductor] patterns to remove pointless view/permute pairs

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136702
* #138429
* #138756
* __->__ #139136

These are not artificial patterns I come up. They shows up in linear+CrossEntropyLoss graph.

Consider this snippet:
```
        class LinearAndCEL(nn.Module):
            def __init__(self):
                super().__init__()
                self.linear = nn.Linear(C, V)
                self.ce = nn.CrossEntropyLoss()

            def forward(self, x, y):
                return self.ce(self.linear(x).view(B * T, V), y.view(-1))
```

`x` passed to `forward` is a 3D tensor of shape [B, T, C].
The `self.linear` will view x as [BxT, C] shape tensor first, do the matmul and produce a [BxT, V] tensor, and then view this output back to a 3D tensor with shape [B, T, V]. User code is gonna add another view op to convert the tensor shape to [B x T, V]. This generates a pair of redundant views . A pair of redundant permute happens in the backward part when we compute gradients.

The view ops makes it hard to chunk linear+CEL. When the view op breaks up the dimension being chunked, what should the chunker do (even if we merge those dimension again later)? Removing these pointless view pairs makes the chunker simpler. And I think it's in general nice to do.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by Chillee — APPROVED** [review]

**Review by jansel — APPROVED** [review]

File: torch/_inductor/fx_passes/joint_graph.py [hunk 1]
```diff
@@ -568,6 +568,45 @@ def pointless_view(match: Match, arg, size):
...
+    ),
+    pass_dict=patterns,
+)
+def pointless_view_pair(match: Match, arg, size1, size2):
+    """
+    Remove a pair of views that are pointless.
+    """
+    node = match.output_node()
+    arg_size = list(arg.meta["val"].shape)
+    if arg_size == size2:
```
Comment (eellison) [inline torch/_inductor/fx_passes/joint_graph.py:585]: This should be using guard_oblivious or we are going to error on unbacked/add guards

File: torch/_inductor/fx_passes/joint_graph.py [hunk 2]
```diff
@@ -568,6 +568,45 @@ def pointless_view(match: Match, arg, size):
...
+    ),
+    pass_dict=patterns,
+)
+def pointless_permute_pair(match: Match, arg, perm1, perm2):
+    rank = len(perm1)
+    assert len(perm2) == rank
+
+    for i in range(rank):
+        if perm1[perm2[i]] != i:
+            return  # bail out
```
Comment (eellison) [inline torch/_inductor/fx_passes/joint_graph.py:604]: This as well

**Review by eellison — COMMENTED** [review]
Should this be part of [remove_noops](https://github.com/pytorch/pytorch/blob/9af18169742c7813dcfcb7497525d2df6644c0a7/torch/_inductor/pattern_matcher.py#L1886) ?

**Comment by shunting314** [comment]
> Should this be part of [remove_noops](https://github.com/pytorch/pytorch/blob/9af18169742c7813dcfcb7497525d2df6644c0a7/torch/_inductor/pattern_matcher.py#L1886) ? 

Looks like remove_noops contains those single op patterns (and it does not use pattern matcher), but the patterns added in this PR involves a pair of ops.

File: torch/_inductor/fx_passes/joint_graph.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/fx_passes/joint_graph.py:604]: Maybe not this one since the items being compared are dimensions rather than size/stride.

**Comment by eellison** [comment]
@shunting314 I more so meant as part of the clean up canonicalization passes we run on generated patterns. but this is pretty rare so not necessary i think.

File: torch/_inductor/fx_passes/joint_graph.py [hunk 3]
```diff
@@ -563,11 +566,50 @@ def pointless_view(match: Match, arg, size):
     """Remove no-op view"""
     node = match.output_node()
     arg_size = list(node.args[0].meta["val"].shape)  # type: ignore[union-attr]
-    if size == arg_size:
+    if guard_size_oblivious(size == arg_size):
```
Comment (eellison) [inline torch/_inductor/fx_passes/joint_graph.py:569]: Hmm, I dont think this is sufficient. The individual elements will still be using the guarding version.  Youd want `len(size) == len(arg_size) and all(guard_size_oblivious(dim1, dim2) for zip(size, arg_size))`

Comment (eellison) [reply torch/_inductor/fx_passes/joint_graph.py:569]: The api is it takes a symint and returns True if it is True without adding additional guards. Here you are calling `List.__eq__` - that's not going to return a symint. It's going to recursively call the normal, non guard size obvlivious eq on all of its elements.

Comment (shunting314) [reply torch/_inductor/fx_passes/joint_graph.py:569]: Got it. Do you want me add an `guard_sizes_oblivious` (plural form) API to `symbolic_shapes.py`? Since I need at least call this in two places (in two patterns).

Comment (eellison) [reply torch/_inductor/fx_passes/joint_graph.py:569]: SGTM - we could also do pytree recursive guard_size_oblivious impl. but sizes sounds good

**Comment by shunting314** [comment]
@ZainRizvi two things that looks very weird:

1. I tried to repro locally but the test pass.
2. From the CI log, the test fail because 'cpp_fused_' string is not found. But it actually appears in the output:
```
2024-10-31T18:51:37.0873122Z     cpp_fused_3(buf20, buf21, buf22)
2024-10-31T18:51:37.0873362Z     # Topologically Sorted Source Nodes: [], Original ATen: []
2024-10-31T18:51:37.0873845Z     torch.ops._c10d_functional.all_reduce_coalesced_.default([buf20, buf21, buf22], 'sum', '0')
2024-10-31T18:51:37.0874080Z     # Topologically Sorted Source Nodes: [], Original ATen: []
```

Not sure if something else goes wrong and it happens to start fail on my PR.

**Comment by ngimel** [comment]
>from the CI log, the test fail because 'cpp_fused_' string is not found. But it actually appears in the output:

That's confusing reporting by FileCheck, but when it's looking for multiple instances and doesn't find the necessary number, it would print search string from the last successful iteration. Try printing and inspecting full `code` string

**Comment by shunting314** [comment]
Thanks for the suggestion @ngimel  :) 

It was hard to debug previously since I can not repro. Now after rebasing, I repro'ed the failure. Will try what you said.

File: test/distributed/_composable/test_replicate_with_compiler.py [hunk 4]
```diff
@@ -318,7 +318,11 @@ def bwd(loss):
     # todo: This pass mucks things up since Inductor thinks its inference
     # and can apply this. Should turn off these passes in compiled autograd
     @torch._inductor.config.patch(
-        reorder_for_locality=False, reorder_for_peak_memory=False
+        reorder_for_locality=False,
+        reorder_for_peak_memory=False,
+        # The correctness of this test relies on the pointless permute ops
+        # in the joint graph does not get eliminated..
+        pattern_matcher=False,
```
Comment (shunting314) [inline test/distributed/_composable/test_replicate_with_compiler.py:325]: cc @fegin

Comment (ngimel) [reply test/distributed/_composable/test_replicate_with_compiler.py:325]: You probably can change what it is checking for, to match newly generated code.

Comment (shunting314) [reply test/distributed/_composable/test_replicate_with_compiler.py:325]: Changing L332 to loop only twice can make the test pass. But I also see the code checks that 'ddp_buckets' equals to 3, so I'm not sure if I'll make the test less helpful by doing that.

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 5]
```diff
@@ -623,10 +623,11 @@ def forward(self, x, y):
                 is_inplace = binary_list[binary_fn][2]
                 # view + linear + view(joint_graph+freeze pass)
                 match_count = match_count + 5 if is_inplace else match_count + 3
-                match_nodes = match_nodes + 7 if is_inplace else match_nodes + 5
+                match_nodes = match_nodes + 8 if is_inplace else match_nodes + 5
```
Comment (huydhn) [inline test/inductor/test_mkldnn_pattern_matcher.py:626]: @shunting314 I think this line is failing in trunk for inductor/test_cpu_cpp_wrapper.py::DynamicShapesCppWrapperCpuTests::test_linear_binary_dynamic_shapes_cpp_wrapper [GH job link](https://github.com/pytorch/pytorch/actions/runs/11672436111/job/32501873847) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/1aa71be56c39908893273bd9558b127159e1ef3a)

It's a slow test that requires `ciflow/slow` to run, so it was missed on your PR.  The failure can be reproduced locally `pytest -v test/inductor/test_cpu_cpp_wrapper.py -k test_linear_binary_dynamic_shapes_cpp_wrapper`
